infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesDeployments.java

```diff
@@ -852,6 +841,31 @@ private String getPodName(String name) throws InfrastructureException {
     return pods.get(0).getMetadata().getName();
   }
 
+  private static class CreateWatcher implements Watcher<Pod> {
+
+    private final CompletableFuture<Pod> future;
+    private final String workspaceId;
+
+    private CreateWatcher(CompletableFuture<Pod> future, String workspaceId) {
+      this.future = future;
+      this.workspaceId = workspaceId;
+    }
+
+    @Override
+    public void eventReceived(Action action, Pod resource) {
+      Map<String, String> labels = resource.getMetadata().getLabels();
+      if (workspaceId.equals(labels.get(CHE_WORKSPACE_ID_LABEL))) {
```

Comment:
I afraid comparing workspace id label only is not enough. It may don't work for multi pods workspaces. Are we able to compare deployment specific label too to ensure that created pod really belongs to needed deployment?

Reply:
I'm not sure there's a good way to do this with the the infra is currently written.

- One path would be to follow `OwnerReferences` on the pod up to the deployment that created it, *a la* `getDeploymentResource()`. This would work but could potentially be pretty slow (multiple calls to OS API).
- Alternatively, we could generate a random label to go with the deployment and pass that in to the watcher, but this seems like a total hack.

The issue I'm having here is that multi-pod/deployment workspaces seem to be impossible on OS at the moment -- the API gives no option for multiple pods in a single deployment -- only one `podSpec` is permitted (although multiple containers in a single pod works perfectly fine), and attempting to make multiple pods/deployments through sequential calls to the `deploy()` method would fail due to a name collision (deployment names are e.g. `workspaceId.dockerfile`). Apart from that, the only unique label assigned to created pods (that we have access to when the watcher is created) is workspaceId. 

In short, AFAICT, the current setup is that we require a 1-1 correspondance between `workspaceId`, a deployment name, and the label given to pods. Unless I'm missing something I feel like this issue should be resolved if/when multi-pod workspaces are implemented.

Reply:
Actually I thought about it more and realized you are right. I'll look into this, and likely add the deployment name as a label in the `deploy()` method.

Reply:
Fixed -- pods now also get a label to refer to the deployment that created them.